Java02. ДЗ 01, Федорова Анна

Fedannie opened the pull request:
HashMap

Fedannie committed:
Java02. ДЗ 01, Федорова Анна

HashMap

sproshev commented on HashMap/.gradle/2.13/taskArtifacts/cache.properties:
```diff
@@ -0,0 +1 @@
+#Sat Sep 17 17:57:25 MSK 2016
```
закоммичено куча лишних файлов

sproshev replied:
git rm file удаляет файл как из репозитория, так и с диска. git rm --cached file, оставляет файл на диске

sproshev commented on HashMap/src/main/java/ru/spbau/java/fedorova/first/HashMap.java:
```diff
@@ -0,0 +1,77 @@
+ */
+public class HashMap {
+    private LinkedList[] list;
+    private int size;
```
очень важное поле и никак не обновляется, -1

sproshev commented on HashMap/src/main/java/ru/spbau/java/fedorova/first/HashMap.java:
```diff
@@ -0,0 +1,77 @@
+ * @author Fedorova Anna
+ */
+public class HashMap {
+    private LinkedList[] list;
```
final

sproshev commented on HashMap/src/main/java/ru/spbau/java/fedorova/first/HashMap.java:
```diff
@@ -0,0 +1,77 @@
+public class HashMap {
+    private LinkedList[] list;
+    private int size;
+    private int capacity = 1000;
```
может константой его сделать?

sproshev commented on HashMap/src/main/java/ru/spbau/java/fedorova/first/HashMap.java:
```diff
@@ -0,0 +1,77 @@
+        return this.size;
+    }
+
+    public int hashCode(String str) {
```
почему паблик?

sproshev commented on HashMap/src/main/java/ru/spbau/java/fedorova/first/HashMap.java:
```diff
@@ -0,0 +1,77 @@
+        if (list[hash].find(key) != null) {
+            return true;
+        }
+        return false;
```
может последние пять строк можно упростить в одну?

sproshev commented on HashMap/src/main/java/ru/spbau/java/fedorova/first/HashMap.java:
```diff
@@ -0,0 +1,77 @@
+    public String get(String key) {
+        int hash = hashCode(key);
+        Node cur_list = list[hash].find(key);
+        if (cur_list == null) {
```
подчеркивания в джаве не используются в именовании

sproshev commented on HashMap/src/main/java/ru/spbau/java/fedorova/first/HashMap.java:
```diff
@@ -0,0 +1,77 @@
+
+    public String put(String key, String value) {
+        int hash = hashCode(key);
+        Node cur_list = list[hash].find(key);
```
может эти две строки выделить в отдельный метод?

sproshev commented on HashMap/src/main/java/ru/spbau/java/fedorova/first/Main.java:
```diff
@@ -0,0 +1,12 @@
+    public static void main(String[] args) {
+        LinkedList ll = new LinkedList();
+    }
+}
```
unused, удалить

sproshev commented on HashMap/src/main/java/ru/spbau/java/fedorova/first/list/LinkedList.java:
```diff
@@ -0,0 +1,90 @@
+        size = 0;
+    }
+
+    public Node getHead() {
```
unused

sproshev replied:
!!!

sproshev commented on HashMap/src/main/java/ru/spbau/java/fedorova/first/list/LinkedList.java:
```diff
@@ -0,0 +1,90 @@
+            return null;
+        }
+        while (cur != null) {
+            if (cur.getKey() == key)
```
точно?

sproshev commented on HashMap/src/main/java/ru/spbau/java/fedorova/first/list/LinkedList.java:
```diff
@@ -0,0 +1,90 @@
+    }
+
+    public void clear() {
+        while (head != tail) {
```
зачем цикл?

sproshev replied:
!!!

sproshev commented on HashMap/src/main/java/ru/spbau/java/fedorova/first/list/LinkedList.java:
```diff
@@ -0,0 +1,90 @@
+    }
+
+    public void print() {
+        System.out.println(this.size);
```
можно обойтись без this

sproshev replied:
и без size тоже можно обойтись, он только здесь очень нужен

sproshev replied:
!!!

sproshev commented on HashMap/src/main/java/ru/spbau/java/fedorova/first/list/node/Node.java:
```diff
@@ -0,0 +1,37 @@
+ * Created on 17.09.2016.
+ * @author Fedorova Anna
+ */
+
```
код не отформатирован, this местами не по делу написан

sproshev commented on HashMap/src/main/java/ru/spbau/java/fedorova/first/list/node/Node.java:
```diff
@@ -0,0 +1,37 @@
+    private String key;
+    private String value;
+    private Node next;
+    public Node(){}
```
unused

sproshev commented on HashMap/src/main/java/ru/spbau/java/fedorova/first/list/node/Node.java:
```diff
@@ -0,0 +1,37 @@
+    }
+    public void setKey(String key) {
+        this.key = key;
+    }
```
unused, и key можно final сделать

sproshev commented on HashMap/src/test/java/ru/spbau/java/fedorova/first/HashMapTest.java:
```diff
@@ -0,0 +1,52 @@
+    public void contains() throws Exception {
+        assertEquals(true, hm.contains("First"));
+        assertEquals(true, hm.contains("Forth"));
+        assertEquals(false, hm.contains("Smth"));
```
assertTrue/False

sproshev commented on HashMap/src/test/java/ru/spbau/java/fedorova/first/HashMapTest.java:
```diff
@@ -0,0 +1,52 @@
+
+    @Test
+    public void get() throws Exception {
+        assertEquals(null, hm.get("Fifth"));
```
assertNull

sproshev commented on HashMap/src/test/java/ru/spbau/java/fedorova/first/HashMapTest.java:
```diff
@@ -0,0 +1,52 @@
+        assertEquals(null, hm.remove("Smth"));
+        assertEquals("1", hm.remove("First"));
+    }
+
```
тесты проверяют не  все (см комментарии), надо обновить

sproshev reviewed (changes requested):
надо исправлять

Fedannie committed:
HashMap 2.0

Исправила.

-----------------------
> +    }
+
+    public int size() {
+        return this.size;
+    }
может последние пять строк можно упростить в одну?

Я не знаю, как их в одну упростить.
----------------

> +                tmp.setNext(tmp.getNext().getNext());
+            }
+        }
точно?

Да вроде, да. Я не вижу ошибки...

sproshev commented on HashMap/src/main/java/ru/spbau/java/fedorova/first/HashMap.java:
```diff
@@ -0,0 +1,79 @@
+public class HashMap {
+    final private LinkedList[] list;
+    private int size = 0;
+    final private int capacity = 1000;
```
константа -- это еще когда static

sproshev commented on HashMap/src/main/java/ru/spbau/java/fedorova/first/HashMap.java:
```diff
@@ -0,0 +1,79 @@
+
+    private boolean findKey(int hash, String key) {
+        return (list[hash].find(key) != null);
+    }
```
используется всего один раз
то, что именуется find*, должно возвращать искомый объект. Т.к. возвращает boolean, то должно быть contains

sproshev replied:
и я имел в виду такой метод:

find(String key) {
int hash = hashCode(key);
return list[hash].find(key);
}

т.к. именно он сокращает дублирование кода, сейчас оно до сих пор есть

sproshev commented on HashMap/src/main/java/ru/spbau/java/fedorova/first/list/LinkedList.java:
```diff
@@ -0,0 +1,90 @@
+    }
+
+    public void clear() {
+        while (head != tail) {
```
!!!

sproshev commented on HashMap/src/main/java/ru/spbau/java/fedorova/first/list/node/Node.java:
```diff
@@ -0,0 +1,37 @@
+        this.next = next;
+    }
+}
+
```
то, что закомментировано, должно быть удалено, оно ж никому не нужно

sproshev commented on HashMap/src/test/java/ru/spbau/java/fedorova/first/list/LinkedListTest.java:
```diff
@@ -0,0 +1,44 @@
+        assertEquals("4", ll.find("Forth").getValue());
+        assertNull(ll.find("First"));
+    }
+}
```
еще в обоих тестах clear можно проверить

sproshev reviewed (approved):
зачтено, 8.5
-0.5 т.к. не все исправлено
